modules/query/src/main/scala/ch/epfl/bluebrain/nexus/admin/query/builder/ResourceRestrictionExpr.scala

```diff
@@ -0,0 +1,47 @@
+package ch.epfl.bluebrain.nexus.admin.query.builder
+
+import ch.epfl.bluebrain.nexus.admin.ld.Const.nxv
+import ch.epfl.bluebrain.nexus.admin.query.filtering.Expr.{ComparisonExpr, LogicalExpr, NoopExpr}
+import ch.epfl.bluebrain.nexus.admin.query.filtering.Op.Or
+import ch.epfl.bluebrain.nexus.admin.query.filtering.PropPath.UriPath
+import ch.epfl.bluebrain.nexus.admin.query.filtering.Term.LiteralTerm
+import ch.epfl.bluebrain.nexus.admin.query.filtering.{Expr, Op}
+import ch.epfl.bluebrain.nexus.admin.refined.project.ProjectReference
+import ch.epfl.bluebrain.nexus.commons.iam.acls.{FullAccessControlList, Path, Permission, Permissions}
+import eu.timepit.refined.api.RefType.applyRef
+import shapeless.{:+:, CNil, Coproduct}
+
+object ResourceRestrictionExpr {
+
+  protected case object Root
+
+  protected type Ids = Root.type :+: ProjectReference :+: CNil
+
+  private implicit def toIds(path: Path): Option[Ids] = path.segments match {
+    case _ :: project :: Nil => applyRef[ProjectReference](project).toOption.map(Coproduct[Ids](_))
+    case _ :: Nil            => Some(Coproduct[Ids](Root))
+    case _                   => None
+  }
+
+  final def apply(acls: FullAccessControlList): Expr = {
```

Comment:
I don't think this would make the code any simpler, because we have to go through all the permissions anyway, because we need to filter the paths where we have `projects/read` and `projects/write`